### ISSUE TITLE
Failure when training with tf.GradientTape() for regression problems

### ISSUE THREAD

**Issue by markodjordjic** [tensorflow/tensorflow#49907]
**System information**
- OS Platform and Distribution: Windows 10 (the same thing happens on Linux)
- TensorFlow version: 2.1.0 (the same thing happens on 2.5.0)
- Python version: 3.6.6

**Describe the current behavior**
For the exactly the same model, training via `tf.GradientTape()` either: (a) does not converge, or (b) converges with worse score than model training via `tf.keras.Model.fit` method. 

**Describe the expected behavior**
Similar training outcome when using `tf.GradinetTape()` and `tf.keras.Model.fit`.

**Standalone code to reproduce the issue**
```
import numpy as np
from sklearn.datasets import load_boston
import tensorflow as tf

tf.random.set_seed(0)


def make_model():
    input_layer = tf.keras.layers.Input(shape=(np.shape(x)[1]))
    inner_layer_1 = tf.keras.layers.Dense(
        units=10,
        activation='selu',
        kernel_initializer=tf.keras.initializers.lecun_normal()
    )(input_layer)
    inner_layer_2 = tf.keras.layers.Dense(
        units=10,
        activation='selu',
        kernel_initializer=tf.keras.initializers.lecun_normal()
    )(inner_layer_1)
    output_layer = tf.keras.layers.Dense(
        units=1,
        activation='linear'
    )(inner_layer_2)
    return tf.keras.Model(
        input_layer,
        output_layer
    )

# Get raw data.
raw_features, y = load_boston(return_X_y=True)

# Standardize features.
x = (raw_features-raw_features.mean(axis=0)) / raw_features.std(axis=0)
print(np.std(x, axis=0))

# Reshape targets.
y = np.array(np.reshape(y, newshape=(-1, 1)), dtype=np.float32)

# Training parameters.
optimizer = tf.keras.optimizers.Adam()
optimizer_mp = tf.keras.mixed_precision.experimental.LossScaleOptimizer(
    optimizer, "dynamic"
)
objective = tf.keras.losses.MeanSquaredError()
batch_size = 4
epochs = 5

# Fit via fit method.
train_fit = make_model()
train_fit.summary()
train_fit.compile(optimizer, objective)
train_fit.fit(x=x, y=y, batch_size=batch_size, epochs=epochs)

# Fit via gradient tape.
gradient_tape_fit = make_model()
gradient_tape_fit.summary()
dataset = tf.data.Dataset.from_tensor_slices((x, y)).batch(batch_size)
for epoch in range(0, epochs):
    for step, (x_batch, y_batch) in enumerate(dataset):
        with tf.GradientTape() as tape:
            predictions = gradient_tape_fit(x_batch, training=True)
            loss = objective(y_batch, predictions)
            scaled_loss = optimizer_mp.get_scaled_loss(loss)
        scaled_grads = tape.gradient(
            scaled_loss, gradient_tape_fit.trainable_weights
        )
        gradients = optimizer_mp.get_unscaled_gradients(
            scaled_grads
        )
        optimizer.apply_gradients(zip(
            gradients, gradient_tape_fit.trainable_weights
        ))
    predictions_via_tape = gradient_tape_fit.predict(x)
    print(
        'Tape MSE: %s' % np.mean(np.power(y-predictions_via_tape, 2))
    )

predictions_via_train = train_fit.predict(x)
print(
    'Fit MSE: %s' % np.mean(np.power(y-predictions_via_train, 2))
)
```

**Comment by tusharvickey1999** [comment]
@markodjordjic the reason could be with way how losses are calculated. Since losses calculated using fit() are scaled during training so it is not exactly the mse but mse with some scaling.

**Comment by markodjordjic** [comment]
@tusharvickey1999 Thank you for your reply. I have updated the implementation, so now there is an optimizer which can retrieve both scaled and unscaled loss. However the issue still remains. There is a significant decline in model quality when trained via `tf.GradientTape()` in comparison to the training wit the  `tf.keras.Model.fit()` method.

**Comment by UsharaniPagadala** [comment]
@markodjordjic 

Please refer similar [issue](https://github.com/tensorflow/tensorflow/issues/33898) and let us know if it helps.Thanks

**Comment by markodjordjic** [comment]
@UsharaniPagadala Hi. Thank you for referring to the issue #33898. The issue deals with the problem in a slightly different way, but it was still possible to serve as an insight in what is the root cause of the problem. And the final conclusion is that the `tf.keras.Model.fit()` method does the **shuffling of the data by default** and then if you wish to replicate the outcome of the training process via `tf.GradientTape()` data also needs to be shuffled before consumed by the training process. This of course refers to training processes where **shuffling of the data is needed**. If shuffling of the data is not needed it **must be disabled in the `tf.keras.Model.fit()`** as well; otherwise the trainings will still have different outcomes.

**Comment by UsharaniPagadala** [comment]
@jvishnuvardhan 
I was able to replicate the issue reported here.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/f03c0ae7ac4b24fd3b57ec8299ef8d87/untitled85.ipynb) here.Thanks
